### PR TITLE
Envest/43 baseline model nb

### LINKED ISSUE #43: Baseline model notebook
This is the baseline of our modeling efforts and aims to compare performance of different methods running.

We have:

- kTSP (with weighting by subgroup size and platform)
- kTSP (without weighting by subgroup size and platform)
- RF (with weighting by subgroup size and platform)
- RF (without weighting by subgroup size and platform)
- LASSO
- MM2S

We will do 10 (or a different number) repeats using the same train/test data across all methods within each repeat.

One repeat will be chosen as the "official" repeat to designate our "official" models, used in testing in other analyses. Options include: median kappa, max kappa. In the past this has been selected randomly.

### PR DESCRIPTION
**Pull request by envest**
Closes #43 

The PR comprises an analysis notebook and the figures it generates for the foundational experiments of this project. Specifically, 

- Can we build a solid single sample predictor (SSP) for medulloblastoma (MB) subgroups?
- Does it work across data platforms?
- Can we improve performance over existing methods, especially in G3/G4 predictions?

### Models

We repeat our experiment 10 times (using different training / testing sets) and build six different models:

1. kTSP (k Top Scoring Pairs) with weighting (weighted)
2. kTSP without weighted (unweighted)
3. RF (Random Forest) (weighted)
4. RF (unweighted)
5. LASSO (we scale the data within sample to make it work like a SSP)
6. MM2S (an existing MB-specific SSP)

Weighting refers to the ability of kTSP and RF models to try to account for imbalances in the size of subgroups and platforms. The hope is that a weighted model allows for better performance of underrepresented subgroups, but it may not. We consider the unweighted kTSP and RF models to be a fairer comparison to the unweighted LASSO and MM2S models.

### The notebook

The notebook takes two parameters:
- `create_models` (default: FALSE) if TRUE, create models (time consuming)
- `overwrite` (default: FALSE) if TRUE, and new models are created, overwrite an existing model file

### Model performance

For each model, we can assess performance using Kappa (an overall measure) and Balanced Accuracy (a subgroup-specific measure). The notebook presents these measures using tables and figures. Balanced Accuracy shows an interesting divergence for G3 samples.

### Visualization of kTSP binary rules

The `multiclassPairs` packages which provides our kTSP and RF modeling methods also provides visualization tools to illustrate model results. The graphic that gets output by their function needs to be captured for integration into a proper manuscript display figure.

The paper figure associated with this notebook combines panels for Kappa, G3 Balanced Accuracy, and the kTSP binary rules. There could be supplemental figures as well, but not at this time.

### Testing:

Project docker image: `envest/mbssp:R-4.2.2`. I will share the model file and the data frame used for plotting with the reviewer separately (on slack).

### Questions I definitely have:

- What are your thoughts on excluding outlier data points to aid visualization? (while specifically mentioning what's weird about them in the text and figure caption)
- Should I track the data used for plotting in `plots/data` or leave it empty? (will plan to share plots/data on figshare if not github)
- What additional information is needed for a new person to read this notebook in the context of the project?
- What notebook tips and tricks would make this notebook easier to work with?

Thank you! 🎉

### CODE REVIEW

**Review by jashapiro — COMMENTED** [review]
This looks good overall. My main overall comment is that it might benefit from just a bit more text at the starts of each section. The sections that have it are good, but it would be nice to have a bit of intro to things like the Kappa and Accuracy tables, as well as an explanation of the metrics being used. Also, you can answer questions like why are "we plotting G3 only?" in that text.

Most of my other comments are style things... 
- Some "modern R" comments like using `\()` for anonymous functions
  - Most of these are really about trying to keep the code within reasonable widths, so there are a good number of reformatting suggestions, as well as piping lists into `purrr`, which is a style choice I prefer, but don't feel too strongly about.
  - A few other little tips, my favorite of which are the `map_vec()` functions to avoid `unlist()` steps and `list_rbind()` because it massively simplifies a kind of gnarly (and brittle?) `melt()` call.
- The other major suggestion is a bit of pulling out parameters in a few places to avoid down-the-road errors.
  - defining the list of model types outside the model-building block so you can reuse it rather than retyping it
  - parameters for variables like replicate and core numbers, and seeds! (variables for sure, so why not parameters)



As noted, I also did a rerun with more replicates (100): As you probably know, the results were pretty consistent, though I didn't see any LASSO failures _quite_ as bad as the one in the data you sent. But the fraction of "failures" was pretty consistent! 11/100 vs 1/10.

Happy to chat about any of my suggestions/changes (note that I pushed `renv`-related changes directly to the branch)!

File: analysis_notebooks/baseline_models.Rmd [hunk 1]
```diff
@@ -456,8 +459,9 @@ For this repeat (only), the model predicts almost all samples to be SHH!
 
 ```{r}
 
-official_model <- which(unlist(purrr::map(baseline_list,
-                                          function(x) x$official_model)))
+official_model <- baseline_list |>
+  purrr::map_lgl(\(x) x$official_model) |>
```
Comment (jashapiro) [inline analysis_notebooks/baseline_models.Rmd:451]: I just remembered that `map` has a really handy extra-shortcut for just this case: 
```suggestion
  purrr::map_lgl("official_model") |>
```

**Comment by envest** [comment]
Implemented / or was inspired to make a change in response to all your comments -- Thanks!

File: renv.lock [hunk 2]
```diff
@@ -2,6 +2,26 @@
...
+        "URL": "https://bioconductor.org/packages/3.16/data/experiment"
+      },
+      {
+        "Name": "BioCworkflows",
+        "URL": "https://bioconductor.org/packages/3.16/workflows"
+      },
+      {
+        "Name": "BioCbooks",
+        "URL": "https://bioconductor.org/packages/3.16/books"
+      },
```
Comment (jashapiro) [inline renv.lock:24]: I don't quite know how these got in here (probably me), but I think I would recommend swapping in RSPM (PPM?), which can allow `renv` to use binary packages on linux (big speed bonus if you want to use the same `renv.lock` file for the docker image)

```suggestion
      {
        "Name": "RSPM",
        "URL": "https://packagemanager.rstudio.com/cran/latest"
      },
```

**Review by jashapiro — APPROVED** [review]
This looks good to me! I do like the additional text that you added.

One additional little thought that I probably should have made earlier when making style suggestions: I am a big fan of the `styler` package, which allows reformatting of R code to a standard style automatically, avoiding things like long lines or inconsistent paren placement. You might consider running it on your code here. (I was inspired by this to suggest adding it as an automatic step in [scpcaTools](https://github.com/AlexsLemonade/scpcaTools/issues/168).

**Pull request merged**